### ISSUE TITLE
[BUG]: test failure on `main`: `ld-floxlib.bats`

### ISSUE THREAD

**Issue by aakropotkin** [flox/flox#862]
**Describe the bug:**

`tests/ld-floxlib.bats` fails on `main` branch for `x86_64-linux` when run interactively.

**Steps to reproduce:**
1. `git checkout main`
2. `git clean -xfd`
3. `nix develop`
4. `just test-all`

Result: 
Segmentation fault while running `ld-floxlib.bats`:

```console
ld-floxlib.bats
 zsh: suspended (tty input)  just test-all
```

Expected: 
Test on `main` should pass

**Screenshots/terminal output (if relevant):**

** flox Version (run `flox --version` if possible): **

**Comment by limeytexan** [comment]
On what Linux distribution / version? Also what is the error - you've just appended the suspended invocation. I successfully ran the test interactively as I was developing the test, and I too saw the suspending thing but figured it was some bats thing - I just typed `fg` three times to complete the test.

**Comment by stahnma** [comment]
I can reproduce this. It's a little odd that I owuld need to do `fg` a few times. How can this be fixed up to work in an automated run?

**Comment by ghudgins** [comment]
@ysndr could use your input on this

**Comment by ghudgins** [comment]
@tomberek @aakropotkin - add notes to this issue

**Comment by tomberek** [comment]
The test suspends/backgrounds and thus does not complete. This causes odd things to happen both locally and in CI. It seems to be related to 
1. using bash
2. to call bats
3. to call run
4. to call bash -exc
https://github.com/flox/flox/blob/main/cli/tests/ld-floxlib.bats#L112

Did not discover root cause yet.

**Comment by aakropotkin** [comment]
Use this branch for reference on _things I tried_.

I was able to isolate the issues down into that script around `realpath`: https://github.com/flox/flox/compare/main...aameen/ld-floxlib/fix

**Comment by limeytexan** [comment]
> I can reproduce this. It's a little odd that I owuld need to do `fg` a few times. How can this be fixed up to work in an automated run?

FTR the CI tests were passing when the PR was merged; the issue of needing to run `fg` only occurred with a controlling TTY at that time.

**Comment by limeytexan** [comment]
Confirmed I don't see any sign of a segfault, but I do see the process group receiving `SIGTTIN`.  

From https://unix.stackexchange.com/questions/550097/a-process-is-being-stopped-by-sigttin-is-there-a-good-way-of-tracing-what-proce

> The SIGTTIN signal is delivered to a background process if it tries to read from the terminal. The default behavior for that signal is to stop the program. ...

The strace also reveals we're invoking `bash -i -c ...` which totally explains it. We can replicate the behaviour with literally any command invoked by bash:

```
flox [flox/prerelease default] [brantley@linuxarm:~/src/flox]$ bash -c true &
[1] 3072558

flox [flox/prerelease default] [brantley@linuxarm:~/src/flox]$ 
[1]+  Done                    bash -c true

flox [flox/prerelease default] [brantley@linuxarm:~/src/flox]$ bash -i -c true &
[1] 3072559

flox [flox/prerelease default] [brantley@linuxarm:~/src/flox]$ 

[1]+  Stopped                 bash -i -c true

flox [flox/prerelease default] [brantley@linuxarm:~/src/flox]$ fg
bash -i -c true

flox [flox/prerelease default] [brantley@linuxarm:~/src/flox]$ 
```

So … we simply cannot do `bash -i` ... in the [old flox-bash](https://github.com/flox/flox-bash/blob/main/lib/commands/activate.sh#L41-L57) we used a Nix-provided bash-noninteractive to source the activation script (using all the modern bash syntax), then we exec'd `$SHELL --norc` (or equivalent args to ensure that the shell invocation doesn't undo all the shell customization performed by the activate). There was a lot of sophisticated logic in there that we don't seem to have brought forward?

**Comment by limeytexan** [comment]
Matthew and I got to the bottom of this:
* the `bash -i` invocation was removed in #889 so that's no longer an issue
* Matthew observed a SIGSEGV being generated on account of `gcc` and `patchelf` tripping over LD_AUDIT in the process of staging the test only, which I suppose is not so surprising. Regardless, those invocations weren't even the "meat" of the test (we could just as easily have committed ELF binaries compiled on a different system), so I don't feel so bad hacking around the problem with the addition of `env -u LD_AUDIT` for those specific command invocations.
* Matthew also found and fixed a small bug with the pkgdb Makefile.

Patch forthcoming presently!

**Comment by limeytexan** [comment]
Addendum: after declaring success (10pm+ on a Friday in London), Matthew observed the CI tests failing way before where they were failing before, right at the point of running _anything_ with LD_AUDIT defined. This would imply ld-floxlib.so is being compiled with a version of glibc that is incompatible with the "system" loader (which by definition tracks the "system" version of glibc). We're digging into what that looks like .. `LD_DEBUG=all` is your friend ...

**Comment by limeytexan** [comment]
Update: we're still having problems running with LD_AUDIT on the GH VM. With Yannik's help I've gained access to connect to the build in progress and find that on the GH Ubuntu 22.04 runner that the use of LD_AUDIT will trigger a segfault when used with _any_ shared library built with the Nix gcc-wrapper (even a trivial/empty one), and that using /usr/bin/cc to compile the very same code works just fine. This has sent me down a deep hole of investigation into the possibility that SELinux, custom GH /etc/ld.so.cache or something/anything else might be triggering these segfaults, and I'm re-learning a ton of things I had forgotten about gdb.

Interesting background on the topic:
* https://web.archive.org/web/20160107032111/http://www.trevorpounds.com/blog/?p=103
* https://www.acrc.bris.ac.uk/acrc/RedHat/rhel-ld-en-4/version.html
* https://www.akkadia.org/drepper/symbol-versioning
* https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-1242.html
* https://man7.org/linux/man-pages/man8/ld.so.8.html (specifically the section in LD_AUDIT which confirms that the application _should_ be loading two copies of libc into the runtime)

Still searching. FWIW/AFAIK we haven't been able to replicate this segfault issue off of the GH runner VMs.

**Comment by limeytexan** [comment]
Managed to get ahold of the core file and look at it with gdb. It's definitely segfaulting writing to the stack:
```
7ffff7dfc8e6 in ?? ()
```

... and here's the relevant slice of the ProcMap:
```
7ffff7dec000-7ffff7f46000 r-xp 00026000 08:01 10109499                   /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/libc.so.6
7ffff7f46000-7ffff7f9b000 r--p 00180000 08:01 10109499                   /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/libc.so.6
...
7ffffffd4000-7ffffffff000 rw-p 00000000 00:00 0                          [stack]
```

This isn't surprising; it's the obvious symptom of _something_ having some notion of a fixed offset into the version of libc that it was built with, but then using that offset with a different version of libc. Digging further with an test program with more symbols with which to figure out what's happening here. At this point I'm confident this is not a problem with our code at all - it's just a problem that this host is not allowing the use of LD_AUDIT with _any_ shared object not built against the "system" version of libc.

**Comment by limeytexan** [comment]
Too much fun; hard to use tools when they themselves are affected by the bug.
```
runner@runner:~$ LD_AUDIT=`find . -name ld-floxlib.so` gdb ./hw
Segmentation fault (core dumped)
runner@runner:~$ ls /nix/store/*/bin/gdb
/nix/store/w9m2bnwl9b45s3vldjk545jv3zxnpx7v-gdb-13.2/bin/gdb
runner@runner:~$ LD_AUDIT=`find . -name ld-floxlib.so` /nix/store/*/bin/gdb  ./hw
GNU gdb (GDB) 13.2
Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-unknown-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ./hw...
(gdb) run
Starting program: /home/runner/hw 
During startup program terminated with signal SIGSEGV, Segmentation fault.
(gdb) bt
No stack.
(gdb) 
```
... which led me to: https://stackoverflow.com/questions/55815912/how-do-i-debug-an-executable-with-gdb-when-it-crashes-during-startup

... which led me to try this:
```
(gdb) catch throw
Catchpoint 1 (throw)
(gdb) break main
Breakpoint 2 at 0x1151: file hw.c, line 4.
(gdb) run
Starting program: /home/runner/hw 
During startup program terminated with signal SIGSEGV, Segmentation fault.
(gdb) 
```

We're back to "dumb" stare and compare experiments. What happens if I take my Ubuntu-compiled binary and hack it to use the Nix loader?
```
runner@runner:~$ patchelf --print-interpreter /nix/store/*/bin/gdb
/nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/ld-linux-x86-64.so.2
runner@runner:~$ patchelf --set-interpreter `!!` ./hw
patchelf --set-interpreter `patchelf --print-interpreter /nix/store/*/bin/gdb` ./hw
runner@runner:~$ ldd ./hw
        linux-vdso.so.1 (0x00007fffe8bc7000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fa64f000000)
        /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x00007fa64f277000)
runner@runner:~$ LD_FLOXLIB_DEBUG=1 !!
LD_FLOXLIB_DEBUG=1 LD_AUDIT=`find . -name ld-floxlib.so` ./hw
DEBUG: la_objsearch(libc.so.6, LA_SER_ORIG)
DEBUG: la_objsearch(/nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/libc.so.6, LA_SER_DEFAULT)
Hello, World!
runner@runner:~$ 
```
So that tells me it's the Ubuntu loader that's segfaulting. Fun.

So time to debug assembly. Wow haven't done that in awhile. Found good article to help with that: https://stackoverflow.com/questions/2420813/using-gdb-to-single-step-assembly-code-outside-specified-executable-causes-error

And ... same problem. Can't get to the point of `stepi`ing a single function.

Going to set this down for awhile. As a last ditch data-gathering effort I gathered an strace invocation along with the corresponding coredump and "hello world" executable (will attach shortly) so that someone with better core dump debugging-fu might be able to discern what's going on here.

**Comment by limeytexan** [comment]
Attempting to upload zip file again ..

**Comment by aakropotkin** [comment]
Linker errors are a tough category of debugging unfortunately.

These logs give me PTSD 😅

I could take a look in a later sprint when the pressure for feature work cools off a bit.

**Comment by limeytexan** [comment]
I've cooked up the following reproducer that fails on the GH runner host and succeeds literally everywhere else I've tried.

1. run `nix develop` to get access to Nix gcc
2. create a completely empty shared object by compiling `/dev/null`
3. run any non-nix executable with LD_AUDIT pointing at the shared object to observe the segfault

Transcript:
```
runner@runner:~/work/flox/flox$ nix develop
warning: Git tree '/home/runner/work/flox/flox' is dirty
runner@runner:~/work/flox/flox$ which gcc
/nix/store/90h6k8ylkgn81k10190v5c9ldyjpzgl9-gcc-wrapper-12.3.0/bin/gcc
runner@runner:~/work/flox/flox$ gcc  -shared -fPIC /dev/null -o issue862.so
runner@runner:~/work/flox/flox$ nm ./issue862.so 
0000000000003df8 d _DYNAMIC
0000000000003fc8 d _GLOBAL_OFFSET_TABLE_
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
0000000000002040 r __FRAME_END__
0000000000004008 d __TMC_END__
                 w __cxa_finalize@GLIBC_2.2.5
00000000000010b0 t __do_global_dtors_aux
0000000000003df0 d __do_global_dtors_aux_fini_array_entry
0000000000004000 d __dso_handle
0000000000003de8 d __frame_dummy_init_array_entry
                 w __gmon_start__
00000000000010fc t _fini
0000000000001000 t _init
0000000000004008 b completed.0
0000000000001040 t deregister_tm_clones
00000000000010f0 t frame_dummy
0000000000001070 t register_tm_clones
runner@runner:~/work/flox/flox$ LD_AUDIT=./issue862.so /usr/bin/true
Segmentation fault (core dumped)
runner@runner:~/work/flox/flox$ 
```

The issue does not happen when compiling with `/usr/bin/cc`:
```
runner@runner:~/work/flox/flox$ /usr/bin/cc  -shared -fPIC /dev/null -o issue862.so                                                                                                           
runner@runner:~/work/flox/flox$ LD_AUDIT=./issue862.so /usr/bin/true
ERROR: ld.so: object './issue862.so' cannot be loaded as audit interface: undefined symbol: la_version; ignored.
runner@runner:~/work/flox/flox$ 
```

The problem seems to be on account of the interpreter being used. I also discovered that the dynamic loader interpreter can be invoked on the command line like you would with any other interpreter (e.g. python, perl). Makes it easier to test than having to use `patchelf --set-interpreter` all the time.  :)
```
runner@runner:~/work/flox/flox$ gcc  -shared -fPIC /dev/null -o issue862.so  # recompile with Nix gcc-wrapper
runner@runner:~/work/flox/flox$ LD_AUDIT=./issue862.so /lib64/ld-linux-x86-64.so.2 /usr/bin/true
Segmentation fault (core dumped)
runner@runner:~/work/flox/flox$ LD_AUDIT=./issue862.so /usr/bin/true
Segmentation fault (core dumped)
runner@runner:~/work/flox/flox$ LD_AUDIT=./issue862.so /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib64/ld-linux-x86-64.so.2 /usr/bin/true
ERROR: ld.so: object './issue862.so' cannot be loaded as audit interface: undefined symbol: la_version; ignored.
runner@runner:~/work/flox/flox$ 
```

Final observation is that the Nix `ldd` always reports a different interpreter than the one compiled into the ELF binary. Related?
```
runner@runner:~/work/flox/flox$ patchelf --print-interpreter /usr/bin/true
/lib64/ld-linux-x86-64.so.2
runner@runner:~/work/flox/flox$ /usr/bin/ldd /usr/bin/true
        linux-vdso.so.1 (0x00007ffff7fc1000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007ffff7c00000)
        /lib64/ld-linux-x86-64.so.2 (0x00007ffff7fc3000)
runner@runner:~/work/flox/flox$ which ldd
/nix/store/cx01qk0qyylvkgisbwc7d3pk8sliccgh-glibc-2.38-27-bin/bin/ldd
runner@runner:~/work/flox/flox$ /nix/store/cx01qk0qyylvkgisbwc7d3pk8sliccgh-glibc-2.38-27-bin/bin/ldd /usr/bin/true
        linux-vdso.so.1 (0x00007ffff7fc8000)
        libc.so.6 => /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib/libc.so.6 (0x00007ffff7dd2000)
        /lib64/ld-linux-x86-64.so.2 => /nix/store/qn3ggz5sf3hkjs2c797xf7nan3amdxmp-glibc-2.38-27/lib64/ld-linux-x86-64.so.2 (0x00007ffff7fca000)
runner@runner:~/work/flox/flox$ 
```

**Comment by limeytexan** [comment]
Got an even better reproducer:
* _very_ simple CI setup that just starts tmate: https://github.com/limeytexan/PR862/blob/main/.github/workflows/ci.yml
* connect to GH runner VM, install Nix with `sh <(curl -L https://nixos.org/nix/install) --daemon`
* install `gcc` to profile with `nix --extra-experimental-features "nix-command flakes" profile install nixpkgs#gcc`
* perform testing similar to above:
```
runner@fv-az1019-782:~/work/PR862/PR862$ /usr/bin/gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

runner@fv-az1019-782:~/work/PR862/PR862$ /usr/bin/gcc  -shared -fPIC /dev/null -o issue862.so
runner@fv-az1019-782:~/work/PR862/PR862$ LD_AUDIT=./issue862.so /usr/bin/true
ERROR: ld.so: object './issue862.so' cannot be loaded as audit interface: undefined symbol: la_version; ignored.
runner@fv-az1019-782:~/work/PR862/PR862$ 
runner@fv-az1019-782:~/work/PR862/PR862$ which gcc
/home/runner/.nix-profile/bin/gcc
runner@fv-az1019-782:~/work/PR862/PR862$ gcc --version
gcc (GCC) 13.2.0
Copyright (C) 2023 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

runner@fv-az1019-782:~/work/PR862/PR862$ gcc  -shared -fPIC /dev/null -o issue862.so
runner@fv-az1019-782:~/work/PR862/PR862$ LD_AUDIT=./issue862.so /usr/bin/true
Segmentation fault (core dumped)
runner@fv-az1019-782:~/work/PR862/PR862$ 
```

**Comment by limeytexan** [comment]
With this reproducer found an interesting difference; running with `LD_DEBUG=bindings LD_AUDIT=./issue862.so /usr/bin/true` reveals on the GH runner that it is reporting "calling init: ./issue862.so" _before_ binding `libc.so.6`, whereas on my local Ubuntu VM it first binds the Nix-provided `libc.so.6` into the `[1]` namespace before "calling init". Digging to see what might cause this change in ordering, or if perhaps it's the reporting of LD_DEBUG diagnostics that is getting re-ordered.

**Comment by limeytexan** [comment]
Worked with Bill a bit to go through the rtld.c code and reason through what must be going on. We identified the "smoking gun" from the output of `LD_DEBUG=bindings`:
```
runner@fv-az1200-29:~/work/PR862/PR862$ LD_DEBUG=bindings LD_AUDIT=./issue862.so /usr/bin/true |& grep -C1 _dl_argv
     30898:     binding file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1] to /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1]: normal symbol `loc1' [GLIBC_2.2.5]
     30898:     binding file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1] to /lib64/ld-linux-x86-64.so.2 [0]: normal symbol `_dl_argv' [GLIBC_PRIVATE]
     30898:     binding file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1] to /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1]: normal symbol `__libc_single_threaded' [GLIBC_2.32]
runner@fv-az1200-29:~/work/PR862/PR862$ 
```

The loader should not be binding the Nix version of libc to the "system" interpreter; note how the lines on either side of that one don't do that, and when we use the Nix-provided interpreter that doesn't happen either.

Since this decision of what file to map to what happens in the interpreter in user space, we're hoping we can instrument a version of the interpreter (or run it in a debugger) to see why it may be making this bad decision. To do this we're downloading and compiling the ubuntu version of glibc with:
* `apt-get source libc6`
* then create a glibc-build directory next to glibc-2.35
* then run `../glibc-2.35/configure --prefix=/usr CC=/usr/bin/gcc CFLAGS="-g -O"`
* then cd into the glibc-build directory
* then run `make CC="/usr/bin/gcc -Wno-maybe-uninitialized"`
* finally, confirm it's broken in user space:
```
runner@fv-az1200-29:~/work/PR862/PR862$ LD_DEBUG=bindings LD_AUDIT=./issue862.so ./glibc-build/elf/ld-linux-x86-64.so.2 /usr/bin/true |& grep -C1 _dl_argv
     52384:     binding file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1] to /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1]: normal symbol `loc1' [GLIBC_2.2.5]
     52384:     binding file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1] to ./glibc-build/elf/ld-linux-x86-64.so.2 [0]: normal symbol `_dl_argv' [GLIBC_PRIVATE]
     52384:     binding file /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1] to /nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib/libc.so.6 [1]: normal symbol `__libc_single_threaded' [GLIBC_2.32]
runner@fv-az1200-29:~/work/PR862/PR862$ 
```

Let the ~games~debugging begin!

**Comment by limeytexan** [comment]
Update:
* compiling and invoking the interpreter in user space does replicate the problem
* running the interpreter from within gdb does not work, as gdb seems to falls victim to the segfault whilst trying to invoke the interpreter(!)
* so we're down to printf debugging
* the tmate session keeps getting torn down at random times, the logs reporting "The hosted runner encountered an error while running your job. (Error Type: Aborted)."
* just before it aborted most recently, I observed that it was attempting to loading `_dl_argv` into the `[1]` namespace (for the audit library) when it should be loaded for the executable itself - note the output should be something more like the following: 
```
LD_DEBUG=all LD_AUDIT=./issue862.so /usr/bin/true |& grep _dl_argv
    579978:     symbol=_dl_argv;  lookup in file=/usr/bin/true [0]
    579978:     symbol=_dl_argv;  lookup in file=/lib/aarch64-linux-gnu/libc.so.6 [0]
    579978:     symbol=_dl_argv;  lookup in file=/lib/ld-linux-aarch64.so.1 [0]
```

Will again re-stage the environment and get back to this point in the debugging.

**Comment by limeytexan** [comment]
Interesting observation; the Nix compiler will force NEEDED libc.so.6 and a runpath:
```
runner@fv-az775-764:~/work/PR862/PR862$ cc  -shared -fPIC /dev/null -o issue862.so
runner@fv-az775-764:~/work/PR862/PR862$ LD_AUDIT=./issue862.so /tmp/true
*** stack smashing detected ***: terminated
Aborted (core dumped)
runner@fv-az775-764:~/work/PR862/PR862$ patchelf --print-needed --print-rpath issue862.so 
/nix/store/cyrrf49i2hm1w7vn2j945ic3rrzgxbqs-glibc-2.38-44/lib:/nix/store/1s98ricsglmfjjqkfnpvywnip5z7gp9q-gcc-13.2.0-lib/lib
libc.so.6
runner@fv-az775-764:~/work/PR862/PR862$ patchelf --remove-rpath issue862.so 
runner@fv-az775-764:~/work/PR862/PR862$ patchelf --remove-needed libc.so.6 issue862.so 
runner@fv-az775-764:~/work/PR862/PR862$ patchelf --print-needed --print-rpath issue862.so 

runner@fv-az775-764:~/work/PR862/PR862$ LD_AUDIT=./issue862.so /tmp/true
Inconsistency detected by ld.so: dl-version.c: 204: _dl_check_map_versions: Assertion `needed != NULL' failed!
runner@fv-az775-764:~/work/PR862/PR862$ 
```

**Comment by limeytexan** [comment]
A script for quickly bootstrapping the env quickly:
```
sh <(curl -L https://nixos.org/nix/install) --daemon --yes
$SHELL -c 'nix --extra-experimental-features "nix-command flakes" profile install nixpkgs#gcc'
$SHELL -c 'gcc  -shared -fPIC /dev/null -o issue862.so'
sudo sed -i 's/# deb-src/deb-src/' /etc/apt/sources.list
sudo apt-get update
apt-get source libc6
git init .
git add glibc-2.35
mkdir glibc-build
cd glibc-build
../glibc-2.35/configure --prefix=/usr CC=/usr/bin/gcc CFLAGS="-g -O"
make CC="/usr/bin/gcc -Wno-maybe-uninitialized -Wno-format-overflow" >& /dev/null
```

Recommend starting a new shell with "ctrl-b c" while that runs, then run the following:
```
ulimit -c unlimited
LD_AUDIT=./issue862.so /usr/bin/true
```

... to be amended/continued ...

**Comment by limeytexan** [comment]
Summarized findings of the past few days:
* The aarch64-linux experience does not match the x86_64-linux one; turns out it's possible to recreate the problem on AWS Amazon Linux and Ubuntu 22.04 instances, but not on the aarch64 one.
    * This is likely related to the fact that aarch64 glibc support seems to have started with version 2.17, while the oldest that x86_64 distributions seem to expect is 2.2.5, and aarch64-linux compilations are binding to 2.17 symbols by default.
* The LD_AUDIT feature as implemented in Linux is hella complicated, and I believe broken with regard to its stated commitment[^1] to maintain isolated namespaces and symbol resolution for each audit library.
    * I wrestled with the task of compiling my own copy of glibc and annotating rtld to map out its function, but the very segfault I was looking to debug plagued all of my attempts to observe and trace the problem.
    * Bottom line is that rtld crashed when attempting to `init` an audit library built against a different version of libc, which to me indicates that it must be mapping the audit library's symbols into the wrong library, thus invalidating offsets compiled in to the (PIC) library.

So in the end, somewhat discouraged at the prospect of ever having a functional LD_AUDIT library I instead turned my attention to the use of `__asm__( ".symver ..." )` statements to compile a library a) with no runpath, and b) requiring only the oldest known versions of the necessary GLIBC functions. I had tried this before, but on this latest attempt I had more experience with the code and was more determined, and _finally_ got it working. *_The tests in #978 are now passing and we'll have this merged shortly_*.

[^1]: [the Oracle (Solaris) documentation on LD_AUDIT](https://docs.oracle.com/cd/E23824_01/html/819-0690/chapter6-1242.html) is far more explicit about this when it says: _"In addition to these two standard link-map lists, the runtime linker allows the creation of an arbitrary number of additional link-map lists. Each of these additional link-map lists provides a unique namespace. The rtld-audit interface employs its own link-map list on which the audit libraries are maintained. The audit libraries are therefore isolated from the symbol binding requirements of the application. Every rtld-audit support library is assigned a unique new link-map identifier."_  The Linux documentation implies the same, although I don't recall at this point where I saw it, and my empirical experience using the rtld might suggest otherwise.

**Comment by aakropotkin** [comment]
I had mentioned this a few months ago, but please be very cautious consulting SunOS docs for Linux DL.

Especially `DT_*` extensiond and all `dlmopen` related calls. GNU devs copied verbatim some SunOS docs for those features but did not align the implementation.
If you dig enough into RHEL bug reports you'll find some for `DT_FILTER` and `dlopen` that discuss this.

**Comment by limeytexan** [comment]
Oh, totally agree. I've spent a ton of time poring over Solaris and Linux documentation and gathering data to compare results of the two. I spent the better part of 2 months dissecting NFS client behaviours on Linux and arguing with RedHat how they needed to be fixed to behave more like their Solaris counterparts. In this respect I view Solaris documentation as more aspirational, and my specific conclusion is that Linux has not yet caught up with where Solaris was with regard to independent namespaces / file binding for LD_AUDIT libraries. (Which is why I'm happy to land on a solution that allows our library to use the libc used by the application itself.)
